File: backend/services/leaderboardService.js
// Serviço de Leaderboard e Sistema de Ranking
class LeaderboardService {
  constructor() {
    // Simulação de banco de dados em memória
    this.players = new Map();
    this.gameStats = new Map();
    this.tournaments = new Map();
    this.seasonalRankings = new Map();
    
    // Inicializar dados de exemplo
    this.initializeExampleData();
  }

  initializeExampleData() {
    // Jogadores de exemplo
    const examplePlayers = [
      { id: '1', name: 'Alexandre, o Grande', level: 25, experience: 15420, avatar: '👑' },
      { id: '2', name: 'Cleópatra VII', level: 22, experience: 12890, avatar: '👸' },
      { id: '3', name: 'Sun Tzu', level: 28, experience: 18750, avatar: '🥋' },
      { id: '4', name: 'Joana d\'Arc', level: 20, experience: 11200, avatar: '⚔️' },
      { id: '5', name: 'Leonardo da Vinci', level: 30, experience: 22100, avatar: '🎨' },
    ];

    examplePlayers.forEach(player => {
      this.players.set(player.id, {
        ...player,
        gamesPlayed: Math.floor(Math.random() * 100) + 50,
        gamesWon: Math.floor(Math.random() * 60) + 20,
        currentStreak: Math.floor(Math.random() * 10),
        bestStreak: Math.floor(Math.random() * 20) + 5,
        favoriteGame: ['Senet', 'Go', 'Mancala', 'Chaturanga'][Math.floor(Math.random() * 4)],
        joinDate: new Date(Date.now() - Math.random() * 365 * 24 * 60 * 60 * 1000),
        lastActive: new Date(Date.now() - Math.random() * 7 * 24 * 60 * 60 * 1000),
        achievements: this.generateRandomAchievements(),
        rating: Math.floor(Math.random() * 1000) + 1200,
      });
    });

    // Estatísticas por jogo
    const games = ['Senet', 'Go', 'Mancala', 'Chaturanga', 'Patolli', 'Hanafuda', 'NineMensMorris', 'Hnefatafl', 'Pachisi'];
    games.forEach(game => {
      this.gameStats.set(game, {
        totalGames: Math.floor(Math.random() * 10000) + 5000,
        totalPlayers: Math.floor(Math.random() * 1000) + 500,
        averageGameTime: Math.floor(Math.random() * 20) + 10, // minutos
        topPlayers: this.generateTopPlayersForGame(game),
      });
    });
  }

  generateRandomAchievements() {
    const allAchievements = [
      'first_win', 'win_streak_5', 'win_streak_10', 'games_played_100',
      'master_senet', 'master_go', 'master_mancala', 'level_10', 'level_20'
    ];
    const count = Math.floor(Math.random() * 5) + 2;
    return allAchievements.slice(0, count);
  }

  generateTopPlayersForGame(game) {
    const playerIds = Array.from(this.players.keys());
    return playerIds
      .sort(() => Math.random() - 0.5)
      .slice(0, 10)
      .map((id, index) => ({
        playerId: id,
        playerName: this.players.get(id).name,
        rating: 1500 - (index * 50) + Math.floor(Math.random() * 100),
        gamesPlayed: Math.floor(Math.random() * 50) + 20,
        winRate: Math.max(0.3, 0.9 - (index * 0.05) + (Math.random() * 0.1)),
      }));
  }

  // Obter ranking global
  getGlobalLeaderboard(limit = 50) {
    const players = Array.from(this.players.values())
      .sort((a, b) => {
        // Ordenar por rating, depois por experiência
        if (b.rating !== a.rating) return b.rating - a.rating;
        return b.experience - a.experience;
      })
      .slice(0, limit)
      .map((player, index) => ({
        rank: index + 1,
        ...player,
        winRate: player.gamesPlayed > 0 ? (player.gamesWon / player.gamesPlayed) : 0,
      }));

    return {
      leaderboard: players,
      totalPlayers: this.players.size,
      lastUpdated: new Date(),
    };
  }

  // Obter ranking por jogo específico
  getGameLeaderboard(gameName, limit = 50) {
    const gameStats = this.gameStats.get(gameName);
    if (!gameStats) {
      throw new Error(`Jogo ${gameName} não encontrado`);
    }

    return {
      game: gameName,
      leaderboard: gameStats.topPlayers.slice(0, limit),
      totalPlayers: gameStats.totalPlayers,
      totalGames: gameStats.totalGames,
      averageGameTime: gameStats.averageGameTime,
      lastUpdated: new Date(),
    };
  }

  // Obter estatísticas de um jogador específico
  getPlayerStats(playerId) {
    const player = this.players.get(playerId);
    if (!player) {
      throw new Error('Jogador não encontrado');
    }

    // Calcular posição no ranking global
    const allPlayers = Array.from(this.players.values())
      .sort((a, b) => b.rating - a.rating);
    const globalRank = allPlayers.findIndex(p => p.id === playerId) + 1;

    // Calcular estatísticas por jogo
    const gameStats = {};
    Array.from(this.gameStats.keys()).forEach(game => {
      const gameLeaderboard = this.getGameLeaderboard(game);
      const playerInGame = gameLeaderboard.leaderboard.find(p => p.playerId === playerId);
      if (playerInGame) {
        gameStats[game] = {
          rank: gameLeaderboard.leaderboard.indexOf(playerInGame) + 1,
          rating: playerInGame.rating,
          gamesPlayed: playerInGame.gamesPlayed,
          winRate: playerInGame.winRate,
        };
      }
    });

    return {
      ...player,
      globalRank,
      gameStats,
      winRate: player.gamesPlayed > 0 ? (player.gamesWon / player.gamesPlayed) : 0,
    };
  }

  // Atualizar estatísticas após uma partida
  updatePlayerAfterGame(playerId, gameResult) {
    const player = this.players.get(playerId);
    if (!player) {
      throw new Error('Jogador não encontrado');
    }

    const { gameName, won, gameTime, opponentRating = 1500 } = gameResult;

    // Atualizar estatísticas básicas
    player.gamesPlayed++;
    if (won) {
      player.gamesWon++;
      player.currentStreak++;
      player.bestStreak = Math.max(player.bestStreak, player.currentStreak);
    } else {
      player.currentStreak = 0;
    }

    // Calcular mudança no rating (sistema ELO simplificado)
    const expectedScore = 1 / (1 + Math.pow(10, (opponentRating - player.rating) / 400));
    const actualScore = won ? 1 : 0;
    const kFactor = 32; // Fator K para mudança de rating
    const ratingChange = Math.round(kFactor * (actualScore - expectedScore));
    
    player.rating = Math.max(800, player.rating + ratingChange);

    // Atualizar experiência
    const baseExp = won ? 100 : 25;
    const timeBonus = Math.max(0, 30 - Math.floor(gameTime / 60)); // Bônus por jogos rápidos
    const streakBonus = player.currentStreak > 1 ? player.currentStreak * 10 : 0;
    
    player.experience += baseExp + timeBonus + streakBonus;

    // Verificar se subiu de nível
    const newLevel = Math.floor(player.experience / 1000) + 1;
    const leveledUp = newLevel > player.level;
    player.level = newLevel;

    // Atualizar última atividade
    player.lastActive = new Date();

    // Verificar conquistas
    const newAchievements = this.checkAchievements(player);

    return {
      player,
      ratingChange,
      experienceGained: baseExp + timeBonus + streakBonus,
      leveledUp,
      newAchievements,
    };
  }

  // Verificar conquistas
  checkAchievements(player) {
    const newAchievements = [];
    const achievements = player.achievements || [];

    // Conquistas de vitórias
    if (player.gamesWon >= 1 && !achievements.includes('first_win')) {
      newAchievements.push('first_win');
    }
    if (player.currentStreak >= 5 && !achievements.includes('win_streak_5')) {
      newAchievements.push('win_streak_5');
    }
    if (player.currentStreak >= 10 && !achievements.includes('win_streak_10')) {
      newAchievements.push('win_streak_10');
    }

    // Conquistas de jogos
    if (player.gamesPlayed >= 100 && !achievements.includes('games_played_100')) {
      newAchievements.push('games_played_100');
    }

    // Conquistas de nível
    if (player.level >= 10 && !achievements.includes('level_10')) {
      newAchievements.push('level_10');
    }
    if (player.level >= 20 && !achievements.includes('level_20')) {
      newAchievements.push('level_20');
    }

    // Adicionar novas conquistas
    player.achievements = [...achievements, ...newAchievements];

    return newAchievements;
  }

  // Criar torneio
  createTournament(tournamentData) {
    const tournamentId = `tournament_${Date.now()}`;
    const tournament = {
      id: tournamentId,
      name: tournamentData.name,
      game: tournamentData.game,
      type: tournamentData.type || 'elimination', // elimination, round_robin, swiss
      maxPlayers: tournamentData.maxPlayers || 16,
      entryFee: tournamentData.entryFee || 0,
      prizePool: tournamentData.prizePool || 0,
      startDate: new Date(tournamentData.startDate),
      endDate: new Date(tournamentData.endDate),
      status: 'registration', // registration, active, completed
      participants: [],
      matches: [],
      createdAt: new Date(),
      createdBy: tournamentData.createdBy,
    };

    this.tournaments.set(tournamentId, tournament);
    return tournament;
  }

  // Obter torneios ativos
  getActiveTournaments() {
    return Array.from(this.tournaments.values())
      .filter(t => t.status === 'registration' || t.status === 'active')
      .sort((a, b) => a.startDate - b.startDate);
  }

  // Inscrever jogador em torneio
  registerPlayerInTournament(tournamentId, playerId) {
    const tournament = this.tournaments.get(tournamentId);
    const player = this.players.get(playerId);

    if (!tournament) throw new Error('Torneio não encontrado');
    if (!player) throw new Error('Jogador não encontrado');
    if (tournament.status !== 'registration') throw new Error('Inscrições encerradas');
    if (tournament.participants.length >= tournament.maxPlayers) throw new Error('Torneio lotado');
    if (tournament.participants.some(p => p.playerId === playerId)) throw new Error('Jogador já inscrito');

    tournament.participants.push({
      playerId,
      playerName: player.name,
      playerRating: player.rating,
      registeredAt: new Date(),
    });

    return tournament;
  }

  // Obter estatísticas gerais da plataforma
  getPlatformStats() {
    const totalPlayers = this.players.size;
    const totalGames = Array.from(this.gameStats.values())
      .reduce((sum, stats) => sum + stats.totalGames, 0);
    
    const activePlayers = Array.from(this.players.values())
      .filter(p => {
        const daysSinceActive = (Date.now() - p.lastActive) / (1000 * 60 * 60 * 24);
        return daysSinceActive <= 7;
      }).length;

    const topGames = Array.from(this.gameStats.entries())
      .sort(([,a], [,b]) => b.totalGames - a.totalGames)
      .slice(0, 5)
      .map(([name, stats]) => ({ name, totalGames: stats.totalGames }));

    return {
      totalPlayers,
      activePlayers,
      totalGames,
      totalTournaments: this.tournaments.size,
      topGames,
      averagePlayerLevel: Array.from(this.players.values())
        .reduce((sum, p) => sum + p.level, 0) / totalPlayers,
    };
  }

  // Obter ranking semanal/mensal
  getSeasonalRanking(season = 'current', game = null) {
    // Implementação simplificada - em produção seria baseado em dados históricos
    const players = Array.from(this.players.values());
    
    if (game) {
      // Filtrar por jogo específico
      const gameStats = this.gameStats.get(game);
      if (gameStats) {
        return {
          season,
          game,
          ranking: gameStats.topPlayers.slice(0, 20),
          totalParticipants: gameStats.totalPlayers,
        };
      }
    }

    // Ranking geral sazonal
    const seasonalPlayers = players
      .map(player => ({
        ...player,
        seasonalRating: player.rating + Math.floor(Math.random() * 100) - 50, // Simulação
        seasonalGames: Math.floor(Math.random() * 20) + 5,
      }))
      .sort((a, b) => b.seasonalRating - a.seasonalRating)
      .slice(0, 50);

    return {
      season,
      ranking: seasonalPlayers,
      totalParticipants: players.length,
      startDate: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000), // 30 dias atrás
      endDate: new Date(),
    };
  }
}

module.exports = LeaderboardService;

File: backend/src/services/leaderboard_service.py
import random
import time
from datetime import datetime, timedelta
from typing import Dict, List, Any, Optional

class LeaderboardService:
    def __init__(self):
        # Simulação de banco de dados em memória
        self.players: Dict[str, Dict] = {}
        self.game_stats: Dict[str, Dict] = {}
        self.tournaments: Dict[str, Dict] = {}
        self.seasonal_rankings: Dict[str, Dict] = {}
        
        # Inicializar dados de exemplo
        self.initialize_example_data()

    def initialize_example_data(self):
        # Jogadores de exemplo
        example_players = [
            {"id": '1', "name": 'Alexandre, o Grande', "level": 25, "experience": 15420, "avatar": '👑'},
            {"id": '2', "name": 'Cleópatra VII', "level": 22, "experience": 12890, "avatar": '👸'},
            {"id": '3', "name": 'Sun Tzu', "level": 28, "experience": 18750, "avatar": '🥋'},
            {"id": '4', "name": 'Joana d\'Arc', "level": 20, "experience": 11200, "avatar": '⚔️'},
            {"id": '5', "name": 'Leonardo da Vinci', "level": 30, "experience": 22100, "avatar": '🎨'},
        ]

        for player in example_players:
            self.players[player["id"]] = {
                **player,
                "gamesPlayed": random.randint(50, 150),
                "gamesWon": random.randint(20, 80),
                "currentStreak": random.randint(0, 10),
                "bestStreak": random.randint(5, 25),
                "favoriteGame": random.choice(['Senet', 'Go', 'Mancala', 'Chaturanga']),
                "joinDate": datetime.now() - timedelta(days=random.randint(1, 365)),
                "lastActive": datetime.now() - timedelta(days=random.randint(0, 7)),
                "achievements": self.generate_random_achievements(),
                "rating": random.randint(1200, 2000),
            }

        # Estatísticas por jogo
        games = ['Senet', 'Go', 'Mancala', 'Chaturanga', 'Patolli', 'Hanafuda', 'NineMensMorris', 'Hnefatafl', 'Pachisi']
        for game in games:
            self.game_stats[game] = {
                "totalGames": random.randint(5000, 15000),
                "totalPlayers": random.randint(500, 1500),
                "averageGameTime": random.randint(10, 30),  # minutos
                "topPlayers": self.generate_top_players_for_game(game),
            }

    def generate_random_achievements(self) -> List[str]:
        all_achievements = [
            'first_win', 'win_streak_5', 'win_streak_10', 'games_played_100',
            'master_senet', 'master_go', 'master_mancala', 'level_10', 'level_20'
        ]
        count = random.randint(2, 7)
        return random.sample(all_achievements, min(count, len(all_achievements)))

    def generate_top_players_for_game(self, game: str) -> List[Dict]:
        player_ids = list(self.players.keys())
        random.shuffle(player_ids)
        
        top_players = []
        for i, player_id in enumerate(player_ids[:10]):
            player = self.players[player_id]
            top_players.append({
                "playerId": player_id,
                "playerName": player["name"],
                "rating": 1500 - (i * 50) + random.randint(-50, 100),
                "gamesPlayed": random.randint(20, 70),
                "winRate": max(0.3, 0.9 - (i * 0.05) + (random.random() * 0.1)),
            })
        
        return sorted(top_players, key=lambda x: x["rating"], reverse=True)

    def get_global_leaderboard(self, limit: int = 50) -> Dict:
        players = list(self.players.values())
        
        # Ordenar por rating, depois por experiência
        players.sort(key=lambda x: (x["rating"], x["experience"]), reverse=True)
        
        leaderboard = []
        for i, player in enumerate(players[:limit]):
            win_rate = player["gamesWon"] / player["gamesPlayed"] if player["gamesPlayed"] > 0 else 0
            leaderboard.append({
                "rank": i + 1,
                **player,
                "winRate": win_rate,
            })

        return {
            "leaderboard": leaderboard,
            "totalPlayers": len(self.players),
            "lastUpdated": datetime.now().isoformat(),
        }

    def get_game_leaderboard(self, game_name: str, limit: int = 50) -> Dict:
        if game_name not in self.game_stats:
            raise ValueError(f"Jogo {game_name} não encontrado")

        game_stats = self.game_stats[game_name]
        
        return {
            "game": game_name,
            "leaderboard": game_stats["topPlayers"][:limit],
            "totalPlayers": game_stats["totalPlayers"],
            "totalGames": game_stats["totalGames"],
            "averageGameTime": game_stats["averageGameTime"],
            "lastUpdated": datetime.now().isoformat(),
        }

    def get_player_stats(self, player_id: str) -> Dict:
        if player_id not in self.players:
            raise ValueError("Jogador não encontrado")

        player = self.players[player_id]

        # Calcular posição no ranking global
        all_players = sorted(self.players.values(), key=lambda x: x["rating"], reverse=True)
        global_rank = next((i + 1 for i, p in enumerate(all_players) if p["id"] == player_id), 0)

        # Calcular estatísticas por jogo
        game_stats = {}
        for game in self.game_stats.keys():
            game_leaderboard = self.get_game_leaderboard(game)
            player_in_game = next((p for p in game_leaderboard["leaderboard"] if p["playerId"] == player_id), None)
            if player_in_game:
                rank = game_leaderboard["leaderboard"].index(player_in_game) + 1
                game_stats[game] = {
                    "rank": rank,
                    "rating": player_in_game["rating"],
                    "gamesPlayed": player_in_game["gamesPlayed"],
                    "winRate": player_in_game["winRate"],
                }

        win_rate = player["gamesWon"] / player["gamesPlayed"] if player["gamesPlayed"] > 0 else 0

        return {
            **player,
            "globalRank": global_rank,
            "gameStats": game_stats,
            "winRate": win_rate,
        }

    def update_player_after_game(self, player_id: str, game_result: Dict) -> Dict:
        if player_id not in self.players:
            raise ValueError("Jogador não encontrado")

        player = self.players[player_id]
        game_name = game_result.get("gameName")
        won = game_result.get("won")
        game_time = game_result.get("gameTime", 900)  # 15 minutos padrão
        opponent_rating = game_result.get("opponentRating", 1500)

        # Atualizar estatísticas básicas
        player["gamesPlayed"] += 1
        if won:
            player["gamesWon"] += 1
            player["currentStreak"] += 1
            player["bestStreak"] = max(player["bestStreak"], player["currentStreak"])
        else:
            player["currentStreak"] = 0

        # Calcular mudança no rating (sistema ELO simplificado)
        expected_score = 1 / (1 + pow(10, (opponent_rating - player["rating"]) / 400))
        actual_score = 1 if won else 0
        k_factor = 32  # Fator K para mudança de rating
        rating_change = round(k_factor * (actual_score - expected_score))
        
        player["rating"] = max(800, player["rating"] + rating_change)

        # Atualizar experiência
        base_exp = 100 if won else 25
        time_bonus = max(0, 30 - (game_time // 60))  # Bônus por jogos rápidos
        streak_bonus = player["currentStreak"] * 10 if player["currentStreak"] > 1 else 0
        
        experience_gained = base_exp + time_bonus + streak_bonus
        player["experience"] += experience_gained

        # Verificar se subiu de nível
        new_level = (player["experience"] // 1000) + 1
        leveled_up = new_level > player["level"]
        player["level"] = new_level

        # Atualizar última atividade
        player["lastActive"] = datetime.now()

        # Verificar conquistas
        new_achievements = self.check_achievements(player)

        return {
            "player": player,
            "ratingChange": rating_change,
            "experienceGained": experience_gained,
            "leveledUp": leveled_up,
            "newAchievements": new_achievements,
        }

    def check_achievements(self, player: Dict) -> List[str]:
        new_achievements = []
        achievements = player.get("achievements", [])

        # Conquistas de vitórias
        if player["gamesWon"] >= 1 and 'first_win' not in achievements:
            new_achievements.append('first_win')
        if player["currentStreak"] >= 5 and 'win_streak_5' not in achievements:
            new_achievements.append('win_streak_5')
        if player["currentStreak"] >= 10 and 'win_streak_10' not in achievements:
            new_achievements.append('win_streak_10')

        # Conquistas de jogos
        if player["gamesPlayed"] >= 100 and 'games_played_100' not in achievements:
            new_achievements.append('games_played_100')

        # Conquistas de nível
        if player["level"] >= 10 and 'level_10' not in achievements:
            new_achievements.append('level_10')
        if player["level"] >= 20 and 'level_20' not in achievements:
            new_achievements.append('level_20')

        # Adicionar novas conquistas
        player["achievements"] = achievements + new_achievements

        return new_achievements

    def create_tournament(self, tournament_data: Dict) -> Dict:
        tournament_id = f"tournament_{int(time.time())}"
        tournament = {
            "id": tournament_id,
            "name": tournament_data["name"],
            "game": tournament_data["game"],
            "type": tournament_data.get("type", "elimination"),
            "maxPlayers": tournament_data.get("maxPlayers", 16),
            "entryFee": tournament_data.get("entryFee", 0),
            "prizePool": tournament_data.get("prizePool", 0),
            "startDate": (datetime.now() + timedelta(days=1)).isoformat(),
            "endDate": (datetime.now() + timedelta(days=7)).isoformat(),
            "status": "registration",
            "participants": [],
            "matches": [],
            "createdAt": datetime.now().isoformat(),
            "createdBy": tournament_data.get("createdBy", "unknown"),
        }

        self.tournaments[tournament_id] = tournament
        return tournament

    def get_active_tournaments(self) -> List[Dict]:
        active_tournaments = []
        for tournament in self.tournaments.values():
            if tournament["status"] in ["registration", "active"]:
                active_tournaments.append(tournament)
        
        return sorted(active_tournaments, key=lambda x: x["startDate"])

    def register_player_in_tournament(self, tournament_id: str, player_id: str) -> Dict:
        if tournament_id not in self.tournaments:
            raise ValueError("Torneio não encontrado")
        if player_id not in self.players:
            raise ValueError("Jogador não encontrado")

        tournament = self.tournaments[tournament_id]
        player = self.players[player_id]

        if tournament["status"] != "registration":
            raise ValueError("Inscrições encerradas")
        if len(tournament["participants"]) >= tournament["maxPlayers"]:
            raise ValueError("Torneio lotado")
        if any(p["playerId"] == player_id for p in tournament["participants"]):
            raise ValueError("Jogador já inscrito")

        tournament["participants"].append({
            "playerId": player_id,
            "playerName": player["name"],
            "playerRating": player["rating"],
            "registeredAt": datetime.now().isoformat(),
        })

        return tournament

    def get_platform_stats(self) -> Dict:
        total_players = len(self.players)
        total_games = sum(stats["totalGames"] for stats in self.game_stats.values())
        
        # Jogadores ativos (últimos 7 dias)
        active_players = sum(1 for player in self.players.values() 
                           if (datetime.now() - player["lastActive"]).days <= 7)

        # Top jogos por número de partidas
        top_games = sorted(
            [(name, stats["totalGames"]) for name, stats in self.game_stats.items()],
            key=lambda x: x[1],
            reverse=True
        )[:5]

        # Nível médio dos jogadores
        average_level = sum(player["level"] for player in self.players.values()) / total_players if total_players > 0 else 0

        return {
            "totalPlayers": total_players,
            "activePlayers": active_players,
            "totalGames": total_games,
            "totalTournaments": len(self.tournaments),
            "topGames": [{"name": name, "totalGames": games} for name, games in top_games],
            "averagePlayerLevel": round(average_level, 1),
        }

    def get_seasonal_ranking(self, season: str = 'current', game: Optional[str] = None) -> Dict:
        # Implementação simplificada - em produção seria baseado em dados históricos
        players = list(self.players.values())
        
        if game and game in self.game_stats:
            # Filtrar por jogo específico
            game_stats = self.game_stats[game]
            return {
                "season": season,
                "game": game,
                "ranking": game_stats["topPlayers"][:20],
                "totalParticipants": game_stats["totalPlayers"],
            }

        # Ranking geral sazonal
        seasonal_players = []
        for player in players:
            seasonal_player = {
                **player,
                "seasonalRating": player["rating"] + random.randint(-50, 50),  # Simulação
                "seasonalGames": random.randint(5, 25),
            }
            seasonal_players.append(seasonal_player)

        seasonal_players.sort(key=lambda x: x["seasonalRating"], reverse=True)

        return {
            "season": season,
            "ranking": seasonal_players[:50],
            "totalParticipants": len(players),
            "startDate": (datetime.now() - timedelta(days=30)).isoformat(),
            "endDate": datetime.now().isoformat(),
        }

